python/lsst/ts/salkafka/component_producer_set.py

```diff
@@ -22,34 +22,75 @@
 
 __all__ = ["ComponentProducerSet"]
 
-
 import argparse
 import asyncio
+import concurrent.futures
+import functools
 import logging
+import multiprocessing
 import os
 import pathlib
+import psutil
+import queue
 import signal
+import sys
+import traceback
 
 from lsst.ts import salobj
-from lsst.ts import salkafka
+from .topic_names_set import TopicNamesSet
+from .component_producer import ComponentProducer
+from .kafka_producer_factory import KafkaConfiguration, KafkaProducerFactory
+
+
+def asyncio_run_method(method, **kargs):
+    """Call asyncio.run on a specified method with specified keyword
+    arguments.
+    """
+    asyncio.run(method(**kargs))
 
 
 class ComponentProducerSet:
     r"""A collection of one or more `ComponentProducer`\ s
     created from a command-line script.
 
+    The normal way to use this class is to run `ComponentProducerSet.amain`
+    from a command-line script. If you wish to run it more directly
+    (e.g. for unit tests), do one of the following:
+
+    * To handle all topics for one or more SAL components::
+
+        kafka_config = KafkaConfig(_kafka info_)
+        producer_set = ComponentProducerSet(kafka_config=kafka_config)
+        await producer_set.run_producers(components=_component names_)
+
+    * To distribute production of one SAL component among multiple
+      subprocesses::
+
+        topic_names_set = TopicNamesSet.from_file(_filepath_)
+        kafka_config = KafkaConfig(_kafka_info_)
+        producer_set = ComponentProducerSet(kafka_config=kafka_config)
+        await producer_set.run_distributed_producer(
+            topic_names_set=topic_names_set,
+        )
+
     Parameters
     ----------
-    args : `argparse.Namespace`
-        Parsed command-line arguments using the argument parser
-        from `make_argument_parser`.
+    kafka_config : `KafkaConfig`
+        Kafka configuration.
+    log_level : `int`, optional
+        Log level, e.g. logging.INFO.
     """
 
-    def __init__(self, args):
+    def __init__(
+        self,
+        kafka_config,
+        log_level=logging.INFO,
+    ):
+        self.kafka_config = kafka_config
 
         self.log = logging.getLogger()
         self.log.addHandler(logging.StreamHandler())
```

Comment:
For some reason this line is causing the logs to be duplicated. I don't understand exactly why that is. Maybe the class is being initialized twice?

In any case, you can add something like this to fix it:

```
if not self.log.hasHandlers():
    self.log.addHandler(logging.StreamHandler())
```

Reply:
I think this needs some attention as well.. double logging in debug mode will be quite cumbersome. One alternative would be to remove this line from here and add it to the initialization method.

Reply:
I'm confused. I already made the change you originally suggested. What else do you think is needed?